# 1d Convolutional Neural Networks

In [0]:

# Imports
import numpy as np
import os
# from utils.utilities import *
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
import tensorflow as tf
import pandas as pd
import tensorflow as tf
from tensorflow import keras

### About the Data

Data is from people doing one of these activities (WALKING, WALKING_UPSTAIRS, WALKING_DOWNSTAIRS, SITTING, STANDING, LAYING) <br><br>
The data is from a smartphone and provides the body acceleration, general acceleration and a gyroscope reading <br><br>
Data is from: https://archive.ics.uci.edu/ml/datasets/human+activity+recognition+using+smartphones 

### Read in data from functions provided below

In [0]:
X_train, labels_train, list_ch_train = read_data_train() # train
X_test, labels_test, list_ch_test = read_data_test()

In [60]:
print(X_train.shape)

(7352, 128, 9)



#### Explaining the shape of the data 
There are 7,352 data points that are classified as doing one of those activities <br>
There are 128 time steps <br>
There are 9 values for each time step for each data point for the x,y,z values of the body 
acceleration, general acceleration and gyrscope reading. <br>

### Use train_test_split to create training and validation data and labels

In [0]:
X_train, X_val, labels_train, labels_val = train_test_split(X_train, labels_train, random_state = 0)

One hot coding is when each row represents one label and it the nonzero column represents what the label is<br>
 [0,0,0,0,0,1] <br>
           ^ This label is LAYING, which was previously represented with a 5
           <br><br>

### One hot encoding functions are provided at the bottom of this notebook <br> Use these functions to convert the training, validation and testing labels

In [0]:
labels_train = one_hot(labels_train)
labels_val = one_hot(labels_val)
labels_test = one_hot(labels_test)

### Define your Keras model <br>
You wlil want to use

See for examples on creating Keras modelshttps://github.com/keras-team/keras/blob/master/examples/mnist_cnn.py

In [77]:
# fill in with your code below
model = keras.Sequential()
model.add(keras.layers.Conv1D(filters=18, kernel_size=3, padding='same', input_shape=X_train[0].shape)) # padding='same' pads so output is same size as input
model.add(keras.layers.GlobalMaxPooling1D())
#model.add(keras.layers.MaxPooling1D(pool_size=2, strides=2))
#model.add(keras.layers.Flatten())
model.add(keras.layers.Dense(labels_train.shape[1], activation=tf.nn.sigmoid)) # sigmoid is good for classification
print(model.summary())  # view model

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv1d_14 (Conv1D)           (None, 128, 18)           504       
_________________________________________________________________
global_max_pooling1d_1 (Glob (None, 18)                0         
_________________________________________________________________
dense_13 (Dense)             (None, 6)                 114       
Total params: 618
Trainable params: 618
Non-trainable params: 0
_________________________________________________________________
None


### Compile your model

In [0]:
lr = 0.001 # choose a learning rate, this will be a good paramter to tune
model.compile( optimizer=tf.train.AdamOptimizer(lr), loss=keras.losses.categorical_crossentropy, metrics = ['accuracy'] )
# categorical crossentropy minimizes the loss/difference between predicted probabilities and real probabilities (one-hot)

### Fit your model

In [79]:
history = model.fit(X_train,
                    labels_train,
                    epochs=100,
                    batch_size=600, # number of data points used in each epoch
                    validation_data=(X_val, labels_val),
                    verbose=1 
                    )

Train on 5514 samples, validate on 1838 samples
Epoch 1/100
5514/5514 [==============================] - 2s 295us/step - loss: 1.7455 - acc: 0.1935 - val_loss: 1.7134 - val_acc: 0.1953
Epoch 2/100
5514/5514 [==============================] - 1s 228us/step - loss: 1.6855 - acc: 0.2031 - val_loss: 1.6601 - val_acc: 0.2100
Epoch 3/100
5514/5514 [==============================] - 1s 227us/step - loss: 1.6339 - acc: 0.2216 - val_loss: 1.6122 - val_acc: 0.2818
Epoch 4/100
5514/5514 [==============================] - 1s 224us/step - loss: 1.5868 - acc: 0.3009 - val_loss: 1.5672 - val_acc: 0.3286
Epoch 5/100
5514/5514 [==============================] - 1s 227us/step - loss: 1.5424 - acc: 0.3422 - val_loss: 1.5241 - val_acc: 0.3743
Epoch 6/100
5514/5514 [==============================] - 1s 228us/step - loss: 1.5000 - acc: 0.3785 - val_loss: 1.4827 - val_acc: 0.3993
Epoch 7/100
5514/5514 [==============================] - 1s 228us/step - loss: 1.4590 - acc: 0.4421 - val_loss: 1.4422 - val_acc: 

Epoch 12/100
5514/5514 [==============================] - 1s 226us/step - loss: 1.2590 - acc: 0.7276 - val_loss: 1.2408 - val_acc: 0.7530
Epoch 13/100
5514/5514 [==============================] - 1s 228us/step - loss: 1.2168 - acc: 0.7583 - val_loss: 1.1984 - val_acc: 0.7856
Epoch 14/100
5514/5514 [==============================] - 1s 225us/step - loss: 1.1735 - acc: 0.7820 - val_loss: 1.1547 - val_acc: 0.8074
Epoch 15/100
5514/5514 [==============================] - 1s 225us/step - loss: 1.1294 - acc: 0.8043 - val_loss: 1.1103 - val_acc: 0.8259
Epoch 16/100
5514/5514 [==============================] - 1s 226us/step - loss: 1.0843 - acc: 0.8210 - val_loss: 1.0651 - val_acc: 0.8373
Epoch 17/100
5514/5514 [==============================] - 1s 223us/step - loss: 1.0391 - acc: 0.8321 - val_loss: 1.0204 - val_acc: 0.8493
Epoch 18/100
5514/5514 [==============================] - 1s 222us/step - loss: 0.9943 - acc: 0.8471 - val_loss: 0.9760 - val_acc: 0.8634
Epoch 19/100
5514/5514 [==========

5514/5514 [==============================] - 1s 228us/step - loss: 0.7879 - acc: 0.8787 - val_loss: 0.7750 - val_acc: 0.8874
Epoch 24/100
5514/5514 [==============================] - 1s 228us/step - loss: 0.7512 - acc: 0.8828 - val_loss: 0.7396 - val_acc: 0.8901
Epoch 25/100
5514/5514 [==============================] - 1s 229us/step - loss: 0.7163 - acc: 0.8867 - val_loss: 0.7060 - val_acc: 0.8928
Epoch 26/100
5514/5514 [==============================] - 1s 226us/step - loss: 0.6831 - acc: 0.8919 - val_loss: 0.6741 - val_acc: 0.8966
Epoch 27/100
5514/5514 [==============================] - 1s 225us/step - loss: 0.6519 - acc: 0.8914 - val_loss: 0.6442 - val_acc: 0.8945
Epoch 28/100
5514/5514 [==============================] - 1s 231us/step - loss: 0.6227 - acc: 0.8948 - val_loss: 0.6166 - val_acc: 0.9004
Epoch 29/100
5514/5514 [==============================] - 1s 225us/step - loss: 0.5958 - acc: 0.8992 - val_loss: 0.5903 - val_acc: 0.9037
Epoch 30/100
5514/5514 [=======================

5514/5514 [==============================] - 1s 225us/step - loss: 0.4892 - acc: 0.9075 - val_loss: 0.4888 - val_acc: 0.9091
Epoch 35/100
5514/5514 [==============================] - 1s 224us/step - loss: 0.4723 - acc: 0.9091 - val_loss: 0.4731 - val_acc: 0.9064
Epoch 36/100
5514/5514 [==============================] - 1s 223us/step - loss: 0.4570 - acc: 0.9086 - val_loss: 0.4585 - val_acc: 0.9108
Epoch 37/100
5514/5514 [==============================] - 1s 226us/step - loss: 0.4425 - acc: 0.9117 - val_loss: 0.4448 - val_acc: 0.9113
Epoch 38/100
5514/5514 [==============================] - 1s 223us/step - loss: 0.4293 - acc: 0.9126 - val_loss: 0.4316 - val_acc: 0.9129
Epoch 39/100
5514/5514 [==============================] - 1s 226us/step - loss: 0.4168 - acc: 0.9131 - val_loss: 0.4201 - val_acc: 0.9119
Epoch 40/100
5514/5514 [==============================] - 1s 225us/step - loss: 0.4050 - acc: 0.9160 - val_loss: 0.4089 - val_acc: 0.9146
Epoch 41/100
5514/5514 [=======================

5514/5514 [==============================] - 1s 225us/step - loss: 0.3553 - acc: 0.9180 - val_loss: 0.3615 - val_acc: 0.9140
Epoch 46/100
5514/5514 [==============================] - 1s 226us/step - loss: 0.3470 - acc: 0.9198 - val_loss: 0.3542 - val_acc: 0.9146
Epoch 47/100
5514/5514 [==============================] - 1s 228us/step - loss: 0.3391 - acc: 0.9184 - val_loss: 0.3458 - val_acc: 0.9135
Epoch 48/100
5514/5514 [==============================] - 1s 226us/step - loss: 0.3316 - acc: 0.9204 - val_loss: 0.3388 - val_acc: 0.9151
Epoch 49/100
5514/5514 [==============================] - 1s 225us/step - loss: 0.3245 - acc: 0.9206 - val_loss: 0.3319 - val_acc: 0.9140
Epoch 50/100
5514/5514 [==============================] - 1s 225us/step - loss: 0.3176 - acc: 0.9218 - val_loss: 0.3254 - val_acc: 0.9157
Epoch 51/100
5514/5514 [==============================] - 1s 224us/step - loss: 0.3112 - acc: 0.9227 - val_loss: 0.3193 - val_acc: 0.9146
Epoch 52/100
5514/5514 [=======================

5514/5514 [==============================] - 1s 226us/step - loss: 0.2832 - acc: 0.9269 - val_loss: 0.2926 - val_acc: 0.9189
Epoch 57/100
5514/5514 [==============================] - 1s 225us/step - loss: 0.2782 - acc: 0.9264 - val_loss: 0.2885 - val_acc: 0.9200
Epoch 58/100
5514/5514 [==============================] - 1s 226us/step - loss: 0.2737 - acc: 0.9267 - val_loss: 0.2837 - val_acc: 0.9238
Epoch 59/100
5514/5514 [==============================] - 1s 226us/step - loss: 0.2691 - acc: 0.9278 - val_loss: 0.2789 - val_acc: 0.9238
Epoch 60/100
5514/5514 [==============================] - 1s 226us/step - loss: 0.2648 - acc: 0.9295 - val_loss: 0.2751 - val_acc: 0.9255
Epoch 61/100
5514/5514 [==============================] - 1s 227us/step - loss: 0.2608 - acc: 0.9280 - val_loss: 0.2713 - val_acc: 0.9227
Epoch 62/100
5514/5514 [==============================] - 1s 225us/step - loss: 0.2567 - acc: 0.9285 - val_loss: 0.2668 - val_acc: 0.9238
Epoch 63/100
5514/5514 [=======================

5514/5514 [==============================] - 1s 224us/step - loss: 0.2387 - acc: 0.9296 - val_loss: 0.2502 - val_acc: 0.9249
Epoch 68/100
5514/5514 [==============================] - 1s 225us/step - loss: 0.2355 - acc: 0.9298 - val_loss: 0.2471 - val_acc: 0.9255
Epoch 69/100
5514/5514 [==============================] - 1s 222us/step - loss: 0.2324 - acc: 0.9300 - val_loss: 0.2437 - val_acc: 0.9249
Epoch 70/100
5514/5514 [==============================] - 1s 226us/step - loss: 0.2296 - acc: 0.9296 - val_loss: 0.2407 - val_acc: 0.9238
Epoch 71/100
5514/5514 [==============================] - 1s 224us/step - loss: 0.2267 - acc: 0.9295 - val_loss: 0.2380 - val_acc: 0.9244
Epoch 72/100
5514/5514 [==============================] - 1s 225us/step - loss: 0.2239 - acc: 0.9300 - val_loss: 0.2355 - val_acc: 0.9244
Epoch 73/100
5514/5514 [==============================] - 1s 228us/step - loss: 0.2212 - acc: 0.9304 - val_loss: 0.2332 - val_acc: 0.9244
Epoch 74/100
5514/5514 [=======================

5514/5514 [==============================] - 1s 226us/step - loss: 0.2092 - acc: 0.9320 - val_loss: 0.2216 - val_acc: 0.9255
Epoch 79/100
5514/5514 [==============================] - 1s 225us/step - loss: 0.2072 - acc: 0.9320 - val_loss: 0.2194 - val_acc: 0.9260
Epoch 80/100
5514/5514 [==============================] - 1s 225us/step - loss: 0.2051 - acc: 0.9314 - val_loss: 0.2174 - val_acc: 0.9260
Epoch 81/100
5514/5514 [==============================] - 1s 226us/step - loss: 0.2031 - acc: 0.9322 - val_loss: 0.2151 - val_acc: 0.9271
Epoch 82/100
5514/5514 [==============================] - 1s 228us/step - loss: 0.2012 - acc: 0.9324 - val_loss: 0.2132 - val_acc: 0.9266
Epoch 83/100
5514/5514 [==============================] - 1s 224us/step - loss: 0.1991 - acc: 0.9324 - val_loss: 0.2117 - val_acc: 0.9271
Epoch 84/100
5514/5514 [==============================] - 1s 226us/step - loss: 0.1974 - acc: 0.9324 - val_loss: 0.2095 - val_acc: 0.9282
Epoch 85/100
5514/5514 [=======================

5514/5514 [==============================] - 1s 228us/step - loss: 0.1896 - acc: 0.9331 - val_loss: 0.2013 - val_acc: 0.9304
Epoch 90/100
5514/5514 [==============================] - 1s 225us/step - loss: 0.1877 - acc: 0.9336 - val_loss: 0.2001 - val_acc: 0.9304
Epoch 91/100
5514/5514 [==============================] - 1s 226us/step - loss: 0.1862 - acc: 0.9336 - val_loss: 0.1984 - val_acc: 0.9298
Epoch 92/100
5514/5514 [==============================] - 1s 227us/step - loss: 0.1847 - acc: 0.9336 - val_loss: 0.1968 - val_acc: 0.9309
Epoch 93/100
5514/5514 [==============================] - 1s 225us/step - loss: 0.1833 - acc: 0.9333 - val_loss: 0.1954 - val_acc: 0.9304
Epoch 94/100
5514/5514 [==============================] - 1s 225us/step - loss: 0.1819 - acc: 0.9336 - val_loss: 0.1940 - val_acc: 0.9309
Epoch 95/100
5514/5514 [==============================] - 1s 225us/step - loss: 0.1806 - acc: 0.9338 - val_loss: 0.1927 - val_acc: 0.9314
Epoch 96/100
5514/5514 [=======================

5514/5514 [==============================] - 1s 226us/step - loss: 0.1743 - acc: 0.9342 - val_loss: 0.1865 - val_acc: 0.9309


### Code below prints the test loss and accuracy <br>
### Before you print your accuracy and loss, play around with the parameters!!

In [80]:
score = model.evaluate(X_test, labels_test, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

('Test loss:', 0.3520313796208376)
('Test accuracy:', 0.8693586698337292)


### Provided Functions Below

In [0]:
def read_data_train():
  """ Read data """

  # Fixed params
  n_class = 6
  n_steps = 128

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/y_train.txt'
  labels = pd.read_csv(label_path, header = None)

  list_of_channels = ['body_acc_x', 'body_acc_y', 'body_acc_z', 'body_gyro_x',
  'body_gyro_y', 'body_gyro_z', 'total_acc_x', 'total_acc_y', 'total_acc_z']

  X = np.zeros((len(labels), n_steps, len(list_of_channels)))

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_x_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,0] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_y_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,1] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_z_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,2] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_x_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,3] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_y_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,4] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_z_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,5] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_x_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,6] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_y_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,7] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_z_train.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,8] = dat_.as_matrix()


  # Return 
  return X, labels[0].values, list_of_channels

In [0]:
def read_data_test():
  """ Read data """

  # Fixed params
  n_class = 6
  n_steps = 128

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/y_test.txt'
  labels = pd.read_csv(label_path, header = None)

  list_of_channels = ['body_acc_x', 'body_acc_y', 'body_acc_z', 'body_gyro_x',
  'body_gyro_y', 'body_gyro_z', 'total_acc_x', 'total_acc_y', 'total_acc_z']

  X = np.zeros((len(labels), n_steps, len(list_of_channels)))

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_x_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,0] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_y_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,1] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_acc_z_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,2] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_x_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,3] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_y_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,4] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/body_gyro_z_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,5] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_x_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,6] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_y_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,7] = dat_.as_matrix()

  label_path = 'https://raw.githubusercontent.com/BeaverWorksMedlytics/datasets/master/week2_conv1d/IntertialSignals/total_acc_z_test.txt'
  dat_ = pd.read_csv(label_path, delim_whitespace = True, header = None)
  X[:,:,8] = dat_.as_matrix()


  # Return 
  return X, labels[0].values, list_of_channels

In [0]:
def one_hot(labels, n_class = 6):
	""" One-hot encoding """
	expansion = np.eye(n_class)
	y = expansion[:, labels-1].T
	assert y.shape[1] == n_class, "Wrong number of labels!"

	return y

In [0]:
def get_batches(X, y, batch_size = 100):
	""" Return a generator for batches """
	n_batches = len(X) // batch_size
	X, y = X[:n_batches*batch_size], y[:n_batches*batch_size]

	# Loop over batches and yield
	for b in range(0, len(X), batch_size):
		yield X[b:b+batch_size], y[b:b+batch_size]